openhands/events/observation/commands.py

```diff
@@ -1,19 +1,136 @@
+        # So that PS1 will keep them as part of the output
+        prompt += json_str.replace('"', r'\"')
+        prompt += CMD_OUTPUT_PS1_END + '\n'  # Ensure there's a newline at the end
+        return prompt
+
+    @classmethod
+    def matches_ps1_metadata(cls, string: str) -> list[re.Match[str]]:
+        matches = []
+        for match in CMD_OUTPUT_METADATA_PS1_REGEX.finditer(string):
+            try:
+                json.loads(match.group(1).strip())  # Try to parse as JSON
+                matches.append(match)
+            except json.JSONDecodeError:
+                logger.warning(
+                    f'Failed to parse PS1 metadata: {match.group(1)}. Skipping.'
+                    + traceback.format_exc()
+                )
+                continue  # Skip if not valid JSON
+        return matches
+
+    @classmethod
+    def from_ps1_match(cls, match: re.Match[str]) -> Self:
+        """Extract the required metadata from a PS1 prompt."""
+        metadata = json.loads(match.group(1))
+        # Create a copy of metadata to avoid modifying the original
+        processed = metadata.copy()
+        # Convert numeric fields
+        if 'pid' in metadata:
+            try:
+                processed['pid'] = int(float(str(metadata['pid'])))
+            except (ValueError, TypeError):
+                processed['pid'] = -1
+        if 'exit_code' in metadata:
+            try:
+                processed['exit_code'] = int(float(str(metadata['exit_code'])))
+            except (ValueError, TypeError):
+                logger.warning(
+                    f'Failed to parse exit code: {metadata["exit_code"]}. Setting to -1.'
+                )
+                processed['exit_code'] = -1
+        return cls(**processed)
+
 
 @dataclass
 class CmdOutputObservation(Observation):
     """This data class represents the output of a command."""
 
-    command_id: int
     command: str
-    exit_code: int = 0
-    hidden: bool = False
     observation: str = ObservationType.RUN
-    interpreter_details: str = ''
+    # Additional metadata captured from PS1
+    metadata: CmdOutputMetadata = field(default_factory=CmdOutputMetadata)
+    # Whether the command output should be hidden from the user
+    hidden: bool = False
+
+    def __init__(
+        self,
+        content: str,
+        command: str,
+        observation: str = ObservationType.RUN,
+        metadata: dict | CmdOutputMetadata | None = None,
+        hidden: bool = False,
+        **kwargs,
+    ):
+        super().__init__(content)
+        self.command = command
+        self.observation = observation
+        self.hidden = hidden
+        if isinstance(metadata, dict):
+            self.metadata = CmdOutputMetadata(**metadata)
+        else:
+            self.metadata = metadata or CmdOutputMetadata()
+
+        # Handle legacy attribute
+        if 'exit_code' in kwargs:
+            self.metadata.exit_code = kwargs['exit_code']
+        if 'command_id' in kwargs:
```

Comment:
Ditto, if we need to handle legacy attrs, we can do it as before, maybe we don't have to write an init (yet)?

Reply:
updated in latest commit!